### ISSUE TITLE
When using Classic Theme, devices not appearing when try to create a threshold

### ISSUE THREAD

**Issue by kamezamw** [Cacti/plugin_thold#368]
When I try to create a threshold, devices are not appearing for both Non Templated or Templated
![image](https://user-images.githubusercontent.com/54975614/64406838-20237a80-d083-11e9-8bd5-e717e0fb34f1.png)

**Comment by DarissAK** [comment]
You can manually tag "&my_host_id=<deviceid>" to the end of the URL to get it past that step. I also had to do that to create a "old" manual threshold because when you select the datasource, the my_host_id is not added to the URL.

**Comment by kamezamw** [comment]
Thanks for the suggestion, it worked to add the threshold but it disappears after a few minutes

**Comment by beernut100** [comment]
I receive the same results. No devices are listed. I can go through and keep adding the new variables to the url until I can create. After it create it seems it disappears during the next scheduled cron run. 


Date | Thu, 26 Sep 2019 09:43:25 -0500
-- | --
Cacti Version | 1.2.6
Cacti OS | unix

Thold version 1.3.2

**Comment by thurban** [comment]
I can confirm this issue with Cacti 1.27 on Windows and Linux.

This is on a newly installed Thold 1.3.2 version.

Devices are not listed, going through the wizard will create a threshold entry, but it ill be deleted (due to the missing host_id entry ... ) with the next poller run.

**Comment by netniV** [comment]
I will see if I can replicate this.

**Comment by thurban** [comment]
The following **line 644** in **thold_webui.php** seems to be odd
			`'sql' => 'SELECT id, description AS name FROM host WHERE disabled!="" AND deleted!=""' . $hiql,`

disabled is set to "on" when a device is disabled, or it is set to "" when not, hence that _where_ statement would select disabled AND deleted devices only ?

```
MariaDB [cacti]> select disabled,deleted from host where id=7;
+----------+---------+
| disabled | deleted |
+----------+---------+
|          |         |
+----------+---------+
1 row in set (0.00 sec)

MariaDB [cacti]> select disabled,deleted from host where id=7;
+----------+---------+
| disabled | deleted |
+----------+---------+
| on       |         |
+----------+---------+
1 row in set (0.00 sec)
```


If i change the "" to "on", then the device shows up, but only when going through the "Create Threshold" button next to a graph.

Creating a "Non-Templated" threshold thorugh Console -> Management -> Thresholds -> Add  will still not list any device

**Comment by netniV** [comment]
So these are non templated items? 

> disabled is set to "on" when a device is disabled, or it is set to "" when not, hence that where statement would select disabled AND deleted devices only ?

I haven't looked at the code yet but I would concur with that assessment.

**Comment by netniV** [comment]
I just created a templated item and it does have the host_id set.  I'll try now with the non-templated style.

**Comment by thurban** [comment]
Yes, non-templated.

If creating Templates first and assign these, they work and also do not get deleted.

**Comment by netniV** [comment]
I have more poller disabled on my dev box, so I'll be able to see before and after a polling :-D

**Comment by netniV** [comment]
So I created a threshold using the + on the thresholds list, selected non-templated and picked a device/datasource.  The resulting insert was:

```
INSERT INTO `thold_data` VALUES (9,'|data_source_description| [|data_source_name|]','localhost - 5 Minute CPU [5m
in_cpu]',40,134,59,0,1,'5min_cpu','','',0,0,'','',1,1,'','',0,0,'','',1,1,0,0,'off',0,0,'','',0,NULL,0,'','0000-00-00 00:00:00','',0,'','',0,0,4,3,0,'',0,0,'','',0,'',0,'','',0,0,'','','','','','','','',0,'',4,3,'','');
```

This shows correct host_id of 4 being inserted.  So, host_id is clearly not our issue here.  I'll take a look at the webui now.

**Comment by netniV** [comment]
```
 select id,host_id from thold_data where id >= 8;
+----+---------+
| id | host_id |
+----+---------+
|  8 |       4 |
|  9 |       4 |
+----+---------+
2 rows in set (0.00 sec)
```
Missed posting this as the proof, 8 is the templated thold, 9 is the untemplated.

**Comment by thurban** [comment]
Here comes the fun part. I am unable to select any device using the process you just described.  It always just has a "Select Device" in there.

**Comment by netniV** [comment]
What were the steps you took?

**Comment by netniV** [comment]
![image](https://user-images.githubusercontent.com/9052188/68590861-6cc27200-0487-11ea-8689-77a23d2c9303.png)
Btw, my UI doesn't show my thresholds for some reason likely what you mentioned.

**Comment by netniV** [comment]
![image](https://user-images.githubusercontent.com/9052188/68591012-adba8680-0487-11ea-83af-1690b029ab93.png)

```
select id,disabled,deleted from host;
+----+----------+---------+
| id | disabled | deleted |
+----+----------+---------+
|  2 |          |         |
|  3 | on       |         |
|  4 |          |         |
|  5 | on       |         |
+----+----------+---------+
4 rows in set (0.00 sec)
```
So, when a device is disabled, it should have ON in the field.  I'll double check the SQL being used.

**Comment by thurban** [comment]
Here's a gif that shows the process/steps:

![cacti_thold_creation](https://user-images.githubusercontent.com/17292964/68591815-de072280-0491-11ea-87fd-079aff67238c.gif)

**Comment by netniV** [comment]
So, if I was you I would change that query to be `disabled=="" and deleted==""` to see if that shows the devices up now.

**Comment by netniV** [comment]
My problem is that an upgrade hasn't added the columns I should have... so I'm investigating why I lost the lastupdated column... I presume you have that one on thold_data?

**Comment by thurban** [comment]
This is from a freshly installed thold 1.3.2:

```
MariaDB [cacti]> describe thold_data;
+-----------------------------+------------------+------+-----+---------------------+----------------+
| Field                       | Type             | Null | Key | Default             | Extra          |
+-----------------------------+------------------+------+-----+---------------------+----------------+
| id                          | int(11)          | NO   | PRI | NULL                | auto_increment |
| name                        | varchar(100)     | YES  |     | NULL                |                |
| name_cache                  | varchar(100)     | YES  |     | NULL                |                |
| local_data_id               | int(11)          | NO   | MUL | 0                   |                |
| data_template_rrd_id        | int(11)          | NO   | MUL | 0                   |                |
| local_graph_id              | int(11)          | NO   | MUL | 0                   |                |
| graph_template_id           | int(11)          | NO   |     | 0                   |                |
| data_template_id            | int(11)          | NO   |     | 0                   |                |
| data_template_hash          | varchar(32)      | NO   |     |                     |                |
| data_source_name            | varchar(100)     | NO   |     |                     |                |
| thold_hi                    | varchar(100)     | YES  |     | NULL                |                |
| thold_low                   | varchar(100)     | YES  |     | NULL                |                |
| thold_fail_trigger          | int(10) unsigned | YES  |     | NULL                |                |
| thold_fail_count            | int(11)          | NO   |     | 0                   |                |
| time_hi                     | varchar(100)     | YES  |     | NULL                |                |
| time_low                    | varchar(100)     | YES  |     | NULL                |                |
| time_fail_trigger           | int(12)          | NO   |     | 1                   |                |
| time_fail_length            | int(12)          | NO   |     | 1                   |                |
| thold_warning_hi            | varchar(100)     | YES  |     | NULL                |                |
| thold_warning_low           | varchar(100)     | YES  |     | NULL                |                |
| thold_warning_fail_trigger  | int(10) unsigned | YES  |     | NULL                |                |
| thold_warning_fail_count    | int(11)          | NO   |     | 0                   |                |
| time_warning_hi             | varchar(100)     | YES  |     | NULL                |                |
| time_warning_low            | varchar(100)     | YES  |     | NULL                |                |
| time_warning_fail_trigger   | int(12)          | NO   |     | 1                   |                |
| time_warning_fail_length    | int(12)          | NO   |     | 1                   |                |
| thold_alert                 | int(1)           | NO   |     | 0                   |                |
| prev_thold_alert            | int(1)           | NO   |     | 0                   |                |
| thold_enabled               | enum('on','off') | NO   | MUL | on                  |                |
| thold_type                  | int(3)           | NO   |     | 0                   |                |
| bl_ref_time_range           | int(10) unsigned | YES  |     | NULL                |                |
| bl_pct_down                 | varchar(100)     | YES  |     | NULL                |                |
| bl_pct_up                   | varchar(100)     | YES  |     | NULL                |                |
| bl_fail_trigger             | int(10) unsigned | YES  |     | NULL                |                |
| bl_fail_count               | int(11) unsigned | YES  |     | NULL                |                |
| bl_alert                    | int(2)           | NO   |     | 0                   |                |
| bl_thold_valid              | int(10) unsigned | NO   |     | 0                   |                |
| lastread                    | varchar(100)     | YES  |     | NULL                |                |
| lasttime                    | timestamp        | NO   |     | 0000-00-00 00:00:00 |                |
| lastchanged                 | timestamp        | NO   |     | 0000-00-00 00:00:00 |                |
| oldvalue                    | varchar(100)     | YES  |     | NULL                |                |
| repeat_alert                | int(10) unsigned | YES  |     | NULL                |                |
| notify_extra                | varchar(512)     | YES  |     | NULL                |                |
| notify_warning_extra        | varchar(512)     | YES  |     | NULL                |                |
| notify_warning              | int(10) unsigned | YES  |     | NULL                |                |
| notify_alert                | int(10) unsigned | YES  |     | NULL                |                |
| snmp_event_category         | varchar(255)     | YES  |     | NULL                |                |
| snmp_event_severity         | tinyint(1)       | NO   |     | 3                   |                |
| snmp_event_warning_severity | tinyint(1)       | NO   |     | 2                   |                |
| thold_daemon_pid            | varchar(25)      | NO   | MUL |                     |                |
| host_id                     | int(10)          | YES  | MUL | NULL                |                |
| syslog_priority             | int(2)           | NO   |     | 3                   |                |
| syslog_facility             | int(2)           | YES  |     | NULL                |                |
| syslog_enabled              | char(3)          | NO   |     |                     |                |
| data_type                   | int(12)          | NO   |     | 0                   |                |
| cdef                        | int(11)          | NO   |     | 0                   |                |
| percent_ds                  | varchar(64)      | NO   |     |                     |                |
| expression                  | varchar(512)     | NO   |     |                     |                |
| thold_template_id           | int(11)          | NO   | MUL | 0                   |                |
| template_enabled            | char(3)          | NO   | MUL |                     |                |
| tcheck                      | int(1)           | NO   | MUL | 0                   |                |
| exempt                      | char(3)          | NO   |     |                     |                |
| acknowledgment              | char(3)          | NO   |     |                     |                |
| thold_hrule_alert           | int(11) unsigned | YES  |     | NULL                |                |
| thold_hrule_warning         | int(11) unsigned | YES  |     | NULL                |                |
| restored_alert              | char(3)          | NO   |     |                     |                |
| reset_ack                   | char(3)          | NO   |     |                     |                |
| persist_ack                 | char(3)          | NO   |     |                     |                |
| email_body                  | varchar(1024)    | NO   |     |                     |                |
| email_body_warn             | varchar(1024)    | NO   |     |                     |                |
| trigger_cmd_high            | varchar(512)     | NO   |     |                     |                |
| trigger_cmd_low             | varchar(512)     | NO   |     |                     |                |
| trigger_cmd_norm            | varchar(512)     | NO   |     |                     |                |
| notes                       | varchar(1024)    | YES  |     |                     |                |
+-----------------------------+------------------+------+-----+---------------------+----------------+

```

**Comment by thurban** [comment]
and no, the device list is still empty when changing the sql query.

**Comment by thurban** [comment]
I wonder why it actually lists devices in your instance, as for the process above, the whole logic is based on a graph_template_id which is not even selected before choosing the device ?

```
} elseif ($type_id == 'thold') {
		$host_template_ids = array_rekey(
			db_fetch_assoc_prepared('SELECT host_template_id
				FROM host_template_graph
				WHERE graph_template_id = ?',
				array($graph_template_id)),
			'host_template_id', 'host_template_id');
		if (cacti_sizeof($host_template_ids)) {
			$hiql = ' AND host_template_id IN (' . implode(', ', $host_template_ids) . ')';
		} else {
			$hiql = ' AND 0 = 1';
		}
		$form_array['my_host_id'] = array(
			'method' => 'drop_callback',
			'friendly_name' => __('Device', 'thold'),
			'description' => __('Select a Device to use for the Threshold and Graph to be created.', 'thold'),
			'on_change' => 'applyTholdFilter()',
			'action' => 'ajax_hosts',
			'id' => $host_id,
			'sql' => 'SELECT id, description AS name FROM host WHERE disabled!="" AND deleted!=""' . $hiql,
			'value' => db_fetch_cell_prepared('SELECT description FROM host WHERE id = ?', array($host_id)),
```

**Comment by thurban** [comment]
so based on the code above, as the graph_tempalte_id is empty, the hiql = ' AND 0 = 1' kicks in, which results in no devices being listed.

What I don't get is that this actually exists since several versions of thold and I'm not sure why this never showed up as an issue before.

**Comment by netniV** [comment]
Strangely that works co me lol

**Comment by thurban** [comment]
Ok,

I changed the lines of code (thold_webapi.php, lines 623 to 647) to this, which solved my problem:

 ```
 } elseif ($type_id == 'thold') {
                $host_template_ids = array_rekey(
                        db_fetch_assoc_prepared('SELECT host_template_id
                                FROM host_template_graph
                                WHERE graph_template_id = ?',
                                array($graph_template_id)),
                        'host_template_id', 'host_template_id');

                if (cacti_sizeof($host_template_ids)) {
                        $hiql = ' AND host_template_id IN (' . implode(', ', $host_template_ids) . ')';
                } else {
                        $hiql = ' AND 1 = 1';
                }

                $form_array['my_host_id'] = array(
                        'method' => 'drop_sql',
                        'friendly_name' => __('Device', 'thold'),
                        'description' => __('Select a Device to use for the Threshold and Graph to be created.', 'thold'),
                        'on_change' => 'applyTholdFilter()',
                        'action' => 'ajax_hosts',
                        'id' => $host_id,
                        'sql' => 'SELECT id, description AS name FROM host WHERE disabled="" AND deleted=""' . $hiql,
                        'value' => $host_id,
                        'none_value' => __('Select a Device', 'thold')
                );
```
So it looks like at least in my instance, the drop_callback doesn't work correctly for that specific form.

**Comment by kamezamw** [comment]
@thurban the lines of code seem to solve my problem as well, so far non-templated tholds have not disappeared.

**Comment by netniV** [comment]
@thurban, can you do a PR for that and I will check it out.

**Comment by thurban** [comment]
I've created the pull request

**Comment by netniV** [comment]
Thanks, having reviewed it and woken up today, I think that is the wrong way to go so I'm going to close it off.  The work around actually creates a security hole in a way, so we are going to have to just narrow down why your devices are not showing.

Can you give me the exact steps you go through to create a threshold?  Can you also let me know if these devices have any data sources/graphs, are they templated in any way (ie, graph or device templates)

**Comment by netniV** [comment]
So, in a separate thread, thanks to @thurban we have identified that this issue seems to be solely for the Classic theme.  It would also appear to be a Cacti Core issue not a thold one per sae with the drop_callback functionality.

I did a review of changes in git for that function:
```
commit 1e2ec65883954f83f16caa5943a2159ad6116763
Author: cigamit <jimmy@sqmail.org>
Date:   Sun Jan 6 18:28:23 2019 -0600

    Form method 'drop_callback' did not respect 'on_change' attribute
```

So it would appear that commit https://github.com/Cacti/cacti/commit/1e2ec65883954f83f16caa5943a2159ad6116763 changed things in between v1.2.0 and v1.2.1 which affects the classic theme.

The above proposed fix is also incorrect as it does not limit the devices to those that the user has visibility of like it should.
